org.eclipse.scanning.device.ui/src/org/eclipse/scanning/device/ui/model/ModelPersistAction.java

```diff
@@ -0,0 +1,152 @@
+	/**
+	 * The action has two types, save and load.
+	 * 
+	 * @author Matthew Gerring
+	 *
+	 */
+	public static enum PersistType {
+		
+		SAVE("Save list to file",   "icons/folder-export.png"), 
+		LOAD("load list from file", "icons/folder-import.png");
+		
+		private final String          label;
+		private final String          descriptor;
+
+		PersistType(String label, String descriptor) {
+		    this.label = label;
+		    this.descriptor = descriptor;
+	    }
+
+		public String getLabel() {
+			return label;
+		}
+
+		public String getDescriptor() {
+			return descriptor;
+		}
+	}
+
+	private final PersistType      type;
+	private final IModelProvider<T> provider;
+	private       Class<T>       modelClass;
+	
+	public ModelPersistAction(IModelProvider<T> provider, PersistType type) {
+		this(provider, type, null);
+	}
+
+	/**
+	 * 
+	 * @param type
+	 */
+	public ModelPersistAction(IModelProvider<T> provider, PersistType type, Class<T> modelClass) {
+		super(type.getLabel(), Activator.getImageDescriptor(type.getDescriptor()));
+		setId(type.toString());
+		this.provider = provider;
+		this.type     = type;
+		this.modelClass = modelClass;
+	}
+	
+	@Override
+	public void run() {
+		
+		try {
+			if (type==PersistType.LOAD) {
+				doLoad();
+			} else if (type==PersistType.SAVE) {
+				doSave();
+			}
+		} catch (Exception neio) {
+			logger.error("Cannot read directory for saved files!", neio);
+		}
+	}
+
+	private void doSave() throws Exception {
+		IStashing stash = chooseFile(type);
+		if (stash==null) return; // They cancelled it
+		stash.save(provider.getModel());
+	}
+
+	private void doLoad() throws IOException {
+		IStashing stash = chooseFile(type);
+		if (stash==null) return; // They cancelled it
+		T path = stash.load(modelClass);
+		try {
+			provider.updateModel(path); // Ensures that validate and events are fired.
+		} catch (Exception e) {
+			logger.error("Problem setting the model. It might be the wrong model type!", e);
+		}
+	}
+
+	private static String previousFile = System.getProperty("GDA/gda.var", System.getProperty("user.home"));
+	private final static String[] extensions = new String[]{"json", "*.*"};
+	private final static String[] files = new String[]{"Model Files (json)", "All Files"};
+	
+	private static final IStashing chooseFile(PersistType type) throws IOException {
+		
+		FileSelectionDialog dialog = new FileSelectionDialog(Display.getCurrent().getActiveShell(), previousFile);
+		dialog.setExtensions(extensions);
+		dialog.setFiles(files);
+		dialog.setNewFile(type==PersistType.SAVE);
+		dialog.setFolderSelector(false);		
+		dialog.create();
+		if (dialog.open() != Dialog.OK) return null;
+		
+		String path = dialog.getPath();
+		if (!path.endsWith(extensions[0])) { //should always be saved to .json
+			path = path.concat("." + extensions[0]);
+		}
+	
+		final File file = new File(path);
+		previousFile    = file.getCanonicalPath();
+		IStashing stash = ServiceHolder.getStashingService().createStash(file);

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "stash". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)